src/hardware/sblaster.cpp

```diff
@@ -396,87 +337,212 @@ static void log_filter_config(const char *output_type, const FilterType filter)
+	uint8_t hpf_order           = {};
+	uint16_t hpf_cutoff_freq_hz = {};
+
+	FilterState lpf_state       = FilterState::Off;
+	uint8_t lpf_order           = {};
+	uint16_t lpf_cutoff_freq_hz = {};
+
+	bool zoh_upsampler_enabled = false;
+	uint16_t zoh_rate_hz       = {};
+};
+
+static void set_filter(mixer_channel_t channel, const FilterConfig &config)
 {
-	auto set_filter = [](const uint8_t order, const uint16_t cutoff_freq) {
-		sb.chan->ConfigureLowPassFilter(order, cutoff_freq);
-		sb.chan->SetLowPassFilter(sb.sb_filter_state);
-	};
+	if (config.hpf_state == FilterState::On ||
+	    config.hpf_state == FilterState::ForcedOn) {
+		channel->ConfigureHighPassFilter(config.hpf_order,
+		                                 config.hpf_cutoff_freq_hz);
+	}
+	channel->SetHighPassFilter(config.hpf_state);
 
-	auto disable_filter = [] { sb.chan->SetLowPassFilter(FilterState::Off); };
+	if (config.lpf_state == FilterState::On ||
+	    config.lpf_state == FilterState::ForcedOn) {
+		channel->ConfigureLowPassFilter(config.lpf_order,
+		                                config.lpf_cutoff_freq_hz);
+	}
+	channel->SetLowPassFilter(config.lpf_state);
 
-	auto enable_zoh_upsampler = [] {
-		constexpr auto dac_rate = 45454;
-		sb.chan->ConfigureZeroOrderHoldUpsampler(dac_rate);
-		sb.chan->EnableZeroOrderHoldUpsampler();
+	if (config.zoh_upsampler_enabled) {
+		channel->ConfigureZeroOrderHoldUpsampler(config.zoh_rate_hz);
+		channel->EnableZeroOrderHoldUpsampler();
+	} else {
+		channel->EnableZeroOrderHoldUpsampler(false);
+	}
+}
+
+static std::optional<FilterType> determine_filter_type(const std::string &filter_choice,
+                                                       const SB_TYPES sb_type)
+{
+	if (filter_choice == "auto") {
+		switch (sb_type) {
+		case SBT_NONE: return FilterType::None;   break;
+		case SBT_1:    return FilterType::SB1;    break;
+		case SBT_2:    return FilterType::SB2;    break;
+		case SBT_PRO1: return FilterType::SBPro1; break;
+		case SBT_PRO2: return FilterType::SBPro2; break;
+		case SBT_16:   return FilterType::SB16;   break;
+		case SBT_GB:   return FilterType::None;   break;
+		}
+	} else if (filter_choice == "off") {
+		return FilterType::None;
+
+	} else if (filter_choice == "sb1") {
+		return FilterType::SB1;
+
+	} else if (filter_choice == "sb2") {
+		return FilterType::SB2;
+
+	} else if (filter_choice == "sbpro1") {
+		return FilterType::SBPro1;
+
+	} else if (filter_choice == "sbpro2") {
+		return FilterType::SBPro2;
+
+	} else if (filter_choice == "sb16") {
+		return FilterType::SB16;
+	}
+
+	return {};
+}
+
+static void configure_sb_filter(mixer_channel_t channel,
+                                const std::string &filter_prefs,
+								const bool filter_always_on,
+                                const SB_TYPES sb_type)
+{
+	// A bit unfortunate, but we need to enable the ZOH upsampler and the the
+	// correct upsample rate first for the filter cutoff frequency validation
+	// to work correctly.
```

Comment:
Makes sense; we need to have our output rate set properly for the filters.

In the channel code, I wonder if we can guarantee this order? For example: is it possible to assert that ZoH is never configured /after/ filter settings have been applied?

Reply:
Yeah it would be better. I'll just put a TODO comment here because I'd rather focus on the `modern` SB filter after this and doing this correctly could be non-trivial (e.g. the init order can get a bit confusing at times).

Reply:
Roger that!